### PR TITLE
protocol/bc: add Entry witness and destination data

### PR DESCRIPTION
**Pull request by bobg**
Add witness and destination data to various entry types in preparation for "entries"-based validation.

A carved-off piece of https://github.com/chain/chain/pull/788

### CODE REVIEW

**Comment by bobg** [comment]
PTAL

**Comment by bobg** [comment]
PTAL

**Comment by kr** [comment]
:eyes:

File: protocol/bc/map.go [hunk 1]
```diff
@@ -68,14 +69,20 @@ func mapTx(tx *TxData) (headerID Hash, hdr *TxHeader, entryMap map[Hash]Entry, e
...
+			)
 
 			if len(oldIss.Nonce) == 0 {
 				if firstSpend == nil {
 					err = fmt.Errorf("nonce-less issuance in transaction with no spends")
 					return
 				}
-				nonce = firstSpend
+				anchor = firstSpend
+				setAnchored = func(id Hash, iss *Issuance) {
```
Comment (kr) [inline protocol/bc/map.go]: This could be `setAnchored = firstSpend.SetAnchored` if you want. Same for nonce below.

File: protocol/bc/value.go [hunk 2]
```diff
@@ -3,7 +3,17 @@ package bc
 type ValueSource struct {
 	Ref      Hash
 	Value    AssetAmount
-	Position uint64 // what int do we actually want?
+	Position uint64
+
+	// The Entry corresponding to Ref, if available
+	// The struct tag excludes the field from hashing
+	Entry `entry:"-"`
```
Comment (kr) [inline protocol/bc/value.go]: Really this comment belongs on ValueDestination.Entry but that line isn't commentable.

Should ValueDestination.Entry have the `entry:"="` tag? It doesn't seem like we ever hash ValueDestination objects.

**Comment by bobg** [comment]
Updated, PTAL

This PR is all that stands in the way of beginning the final "entries validation" code review.

**Comment by kr** [comment]
:eyes:

**Comment by kr** [comment]
LGTM

**Pull request merged**